### PR TITLE
Remove trappy timeouts in snapshot APIs

### PR DESCRIPTION
**Pull request by DaveCTurner**
Wholesale fix of every `TRAPPY_IMPLICIT_DEFAULT_MASTER_NODE_TIMEOUT` in
`o.e.snapshots` and `o.e.repositories`, just pulling them up to the REST
layer (where they become API params), the test suite (where they become
`TEST_REQUEST_TIMEOUT`), or some other place where an explicit value is
available.

Relates #107984

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/cleanup/CleanupRepositoryRequest.java [hunk 1]
```diff
@@ -20,18 +22,34 @@ public class CleanupRepositoryRequest extends AcknowledgedRequest<CleanupReposit
...
-    public CleanupRepositoryRequest(String repository) {
-        super(TRAPPY_IMPLICIT_DEFAULT_MASTER_NODE_TIMEOUT, DEFAULT_ACK_TIMEOUT);
+    public CleanupRepositoryRequest(TimeValue masterNodeTimeout, TimeValue ackTimeout, String repository) {
+        super(masterNodeTimeout, ackTimeout);
         this.repository = repository;
     }
 
-    public CleanupRepositoryRequest(StreamInput in) throws IOException {
-        super(TRAPPY_IMPLICIT_DEFAULT_MASTER_NODE_TIMEOUT, DEFAULT_ACK_TIMEOUT);
+    public static CleanupRepositoryRequest readFrom(StreamInput in) throws IOException {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/cleanup/CleanupRepositoryRequest.java:30]: This request class was missing the `AcknowledgedRequest` header on the wire.

File: server/src/main/java/org/elasticsearch/indices/recovery/plan/ShardSnapshotsService.java [hunk 2]
```diff
@@ -91,7 +93,13 @@ public void fetchLatestSnapshotsForShard(ShardId shardId, ActionListener<Optiona
 
         logger.debug("Searching for peer recovery compatible snapshots in [{}]", repositories);
 
-        GetShardSnapshotRequest request = GetShardSnapshotRequest.latestSnapshotInRepositories(shardId, repositories);
+        GetShardSnapshotRequest request = GetShardSnapshotRequest.latestSnapshotInRepositories(
+            clusterService.state().getMinTransportVersion().onOrAfter(TransportVersions.SNAPSHOT_REQUEST_TIMEOUTS)
+                ? TimeValue.MINUS_ONE
+                : TimeValue.MAX_VALUE,
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/indices/recovery/plan/ShardSnapshotsService.java:99]: This is an internal request so really should have infinite timeout (as long as the cluster is new enough to understand that - using the current PR's transport version to be sure). Previously it had a 30s timeout which was probably a mistake.

File: server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestResetFeatureStateAction.java [hunk 3]
```diff
@@ -43,8 +44,7 @@ public String getName() {
 
     @Override
     protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
-        final ResetFeatureStateRequest req = new ResetFeatureStateRequest();
-
+        final var req = new ResetFeatureStateRequest(RestUtils.getMasterNodeTimeout(request));
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestResetFeatureStateAction.java:47]: This API param was missing - added here and in the JSON spec.

File: server/src/main/java/org/elasticsearch/rest/action/cat/RestSnapshotAction.java [hunk 4]
```diff
@@ -49,16 +49,18 @@ public String getName() {
         return "cat_snapshot_action";
     }
 
+    private static final String[] MATCH_ALL_PATTERNS = { ResolvedRepositories.ALL_PATTERN };
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/rest/action/cat/RestSnapshotAction.java:52]: Pulled this up to a static constant.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/WaitForSnapshotStep.java [hunk 5]
```diff
@@ -104,7 +104,7 @@ public void evaluateCondition(Metadata metadata, Index index, Listener listener,
         );
         String snapshotName = snapPolicyMeta.getLastSuccess().getSnapshotName();
         String repositoryName = snapPolicyMeta.getPolicy().getRepository();
-        GetSnapshotsRequest request = new GetSnapshotsRequest().repositories(repositoryName)
+        GetSnapshotsRequest request = new GetSnapshotsRequest(TimeValue.MAX_VALUE).repositories(repositoryName)
```
Comment (DaveCTurner) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/WaitForSnapshotStep.java:107]: This timeout was missing previously (i.e. defaulted to 30s) but like other ILM actions it should have been `MAX_VALUE`.

**Review by DaveCTurner — COMMENTED** [review]
Big change in terms of lines touched but it was all pretty mechanical apart from the few spots I've called out below.

**Review by ywangd — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/create/CreateSnapshotRequest.java [hunk 6]
```diff
@@ -81,8 +82,8 @@ public class CreateSnapshotRequest extends MasterNodeRequest<CreateSnapshotReque
     @Nullable
     private Map<String, Object> userMetadata;
 
-    public CreateSnapshotRequest() {
-        super(TRAPPY_IMPLICIT_DEFAULT_MASTER_NODE_TIMEOUT);
+    public CreateSnapshotRequest(TimeValue masterNodeTimeout) {
+        super(masterNodeTimeout);
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/create/CreateSnapshotRequest.java:86]: They don't (directly) involve waiting for other nodes to apply the cluster state.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/cleanup/CleanupRepositoryRequest.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/cleanup/CleanupRepositoryRequest.java:30]: No need I think, we'll be clearing up the transport versions in v9 anyway and that'll expose cases like this.

File: server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestResetFeatureStateAction.java (hunk 3 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestResetFeatureStateAction.java:47]: Ah I didn't even realise this had docs. It's not really intended for end-users:

https://github.com/elastic/elasticsearch/blob/5e816686c5d2659a4ebf779dc69e0c5b1f91800b/docs/reference/features/apis/reset-features-api.asciidoc#L11

File: server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/restore/RestoreSnapshotRequest.java [hunk 7]
```diff
@@ -63,8 +64,13 @@ public class RestoreSnapshotRequest extends MasterNodeRequest<RestoreSnapshotReq
...
-    public RestoreSnapshotRequest() {
-        super(TRAPPY_IMPLICIT_DEFAULT_MASTER_NODE_TIMEOUT);
+    public RestoreSnapshotRequest(TimeValue masterNodeTimeout) {
+        super(masterNodeTimeout);
+    }
+
+    @Deprecated(forRemoval = true) // temporary compatibility shim
+    public RestoreSnapshotRequest(String repository, String snapshot) {
+        this(MasterNodeRequest.TRAPPY_IMPLICIT_DEFAULT_MASTER_NODE_TIMEOUT, repository, snapshot);
     }
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/restore/RestoreSnapshotRequest.java:74]: Yep

**Pull request merged**